### DISCUSSION TITLE
System Clock Configuration for STM32MP1

### DISCUSSION THREAD

**Discussion by bhavikbhansali-eaton** [zephyrproject-rtos/zephyr#51933] [Q&A]
We are trying to develop a function which is very much similar to what STM32 Cube IDE generate for system clock configuration. We are using zephyr and I want to add that into our project. We are using STM32MP1 series controller for M4 controller and that is without using HAL of ST. If anyone has done this or have any idea apricate your input. Also I see that in engineering mode M4 is works with 64MHz and that is what I am trying to modify this and take it to 209MHz in engineering mode too.

**Reference:**
Reference to Function, similar to which we are trying to develop with zephyr https://github.com/eziya/STM32_SPI_MCP2515/blob/master/Src/main.c#L148
(I am looking to have function like SystemClock_Config(void) in above mentioned link)

@FRASTM @erwango

**Comment by FRASTM** [comment]
The [clock_stm32_ll_mp1.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/clock_control/clock_stm32_ll_mp1.c)  might be a good starting point to be completed

**Comment by erwango** [comment]
In usual case, we expect A9 core to set up M4 clock. Are you sure you want to deviate from this use case ?
If you want to configure M4 core on its own, you will need to take care there is no conflit from Linux side.
@arnopo Are you aware of similar work ?

**Reply by bhavikbhansali-eaton** [reply]
Yes. That is the actual purpose where we need to configure the M4 on its own.
You have any other solution ?

**Comment by arnopo** [comment]
Right the clock tree is managed by Linux on the Cortex-A9. The Cortex-M4 has just to enable/disable peripheral clock to maintain the clock.
In addition, on Cortex-A9 there is a resource manager to declare peripherals used by the Cortex-M4.
This wiki page should help you: https://wiki.st.com/stm32mpu/wiki/How_to_configure_system_resources
Here is an example of the DTS file associated: https://github.com/STMicroelectronics/linux/blob/v5.15-stm32mp/arch/arm/boot/dts/stm32mp157f-dk2-m4-examples.dts
